### PR TITLE
feat: add plugin config

### LINKED ISSUE #3532: request help: what's the plan of plugin reference
I remember we have discussions about the plugin reference (by adding plugin_id in Route), what's the plan to do this?

### PR DESCRIPTION
**Pull request by spacewander**
Fix #3532
Signed-off-by: spacewander <spacewanderlzx@gmail.com>

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/cli/etcd.lua [hunk 1]
```diff
@@ -251,7 +251,7 @@ function _M.init(env, args)
         for _, dir_name in ipairs({"/routes", "/upstreams", "/services",
                                    "/plugins", "/consumers", "/node_status",
                                    "/ssl", "/global_rules", "/stream_routes",
-                                   "/proto", "/plugin_metadata"}) do
+                                   "/proto", "/plugin_metadata", "/plugin_configs"}) do
```
Comment (membphis) [inline apisix/cli/etcd.lua:254]: Is it possible to move this logic to the `init_by_lua phase` in the future?

File: apisix/init.lua [hunk 2]
```diff
@@ -381,6 +383,18 @@ function _M.http_access_phase()
                   core.json.delay_encode(api_ctx.matched_route, true))
 
     local enable_websocket = route.value.enable_websocket
+
+    if route.value.plugin_config_id then
+        local pc = plugin_config.get(route.value.plugin_config_id)
```
Comment (membphis) [inline apisix/init.lua:388]: `pc` is not a good name

File: apisix/plugin_config.lua [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+end
+
+
+function _M.get(id)
+    return plugin_configs:get(id)
+end
+
+
+function _M.merge(route_conf, plugin_config)
+    if route_conf.prev_plugin_config_ver ~= plugin_config.modifiedIndex then
```
Comment (membphis) [inline apisix/plugin_config.lua:48]: another code style:

```lua
if route_conf.prev_plugin_config_ver == plugin_config.modifiedIndex then
    return route_conf
end

...
```

File: apisix/schema_def.lua [hunk 4]
```diff
@@ -746,6 +749,20 @@ _M.plugins = {
 }
 
 
+_M.plugin_config = {
+    type = "object",
+    properties = {
+        id = id_schema,
+        desc = {type = "string", maxLength = 256},
```
Comment (membphis) [inline apisix/schema_def.lua:756]: all of the `desc` should be the same, declare it as a variable and just quote it

**Review by membphis — APPROVED** [review]
LGTM

File: apisix/schema_def.lua (hunk 4 above)
Comment (spacewander) [reply apisix/schema_def.lua:756]: Will do it in the next PR.

File: apisix/cli/etcd.lua (hunk 1 above)
Comment (spacewander) [reply apisix/cli/etcd.lua:254]: After we can sync etcd data in that phase.

File: apisix/admin/plugin_config.lua [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    return res.status, res.body
+end
+
+
+function _M.delete(id)
+    local key = "/plugin_configs/" .. id
+    local res, err = core.etcd.delete(key)
+    if not res then
+        core.log.error("failed to delete global rule[", key, "]: ", err)
```
Comment (tokers) [inline apisix/admin/plugin_config.lua:105]: Typo: should be "failed to delete plugin_config"

File: doc/architecture-design.md [hunk 6]
```diff
@@ -614,6 +615,129 @@ curl https://{apisix_listen_address}/apisix/admin/global_rules
 
 [Back to top](#table-of-contents)
 
+## Plugin Config
+
+To reuse common plugin configurations, you can exact them into a plugin config and
```
Comment (tokers) [inline doc/architecture-design.md:620]: `exact` should be `extract`?

File: t/admin/plugin-configs.t [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+            local prev_create_time = res.body.node.value.create_time
+            assert(prev_create_time ~= nil, "create_time is nil")
+            local prev_update_time = res.body.node.value.update_time
+            assert(prev_update_time ~= nil, "update_time is nil")
+            ngx.sleep(1)
+
+            local code, body = t('/apisix/admin/plugin_configs/1/plugins',
+                ngx.HTTP_PATCH,
+                [[{
+                    "limit-count": {
```
Comment (tokers) [inline t/admin/plugin-configs.t:243]: Should use some different parameters for `limit-count` since the value is totally same as the one in test case 4.

Comment (spacewander) [reply t/admin/plugin-configs.t:243]: @tokers 
Updated

**Review by tokers — APPROVED** [review]

**Comment by membphis** [comment]
merged, many thx

**Pull request merged**